channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::BTreeMap;
+
+use actix_web::{http, HttpRequest};
+use maxminddb::{self, geoip2::City};
+
+use session::WsChannelSessionState;
+
+// Sender meta data, drawn from the HTTP Headers of the connection counterpart.
+#[derive(Serialize, Debug, Clone)]
+pub struct SenderData {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ua: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub addr: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub city: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub region: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub country: Option<String>,
+}
+
+impl Default for SenderData {
+    fn default() -> Self {
+        Self {
+            ua: None,
+            addr: None,
+            city: None,
+            region: None,
+            country: None,
+        }
+    }
+}
+
+// Parse the Accept-Language header to get the list of preferred languages.
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
+    let default_lang = String::from("en");
+    let mut lang_tree: BTreeMap<String, String> = BTreeMap::new();
+    let mut i = 0;
+    alheader.split(",").for_each(|l| {
+        if l.contains(";") {
+            let weight: Vec<&str> = l.split(";").collect();
+            let lang = weight[0].to_ascii_lowercase();
+            let pref = weight[1].to_ascii_lowercase();
```

Comment:
Should this strip whitespace from `lang` and `pref`?

Reply:
yes.

Reply:
I'm adding a `.trim()` on the insertion call.